src/libraries/Common/src/System/Drawing/ColorTable.cs

```diff
@@ -19,7 +20,9 @@ internal static class ColorTable
             return colors;
         }
 
-        private static void FillWithProperties(Dictionary<string, Color> dictionary, Type typeWithColors)
+        private static void FillWithProperties(
+            Dictionary<string, Color> dictionary,
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties | DynamicallyAccessedMemberTypes.NonPublicProperties)] Type typeWithColors) // can remove NonPublicProperties once https://github.com/mono/linker/issues/1724 is fixed
```

Comment:
Out of curiosity, have we discussed the linker inferring such attribution? i.e. it can see when processing this method that typeWithColors.GetProperties(Public | ...) is used, and it could add DynamicallyAccessedMembers automatically to the metadata if there wasn't already such an attribute.

Reply:
Yes - we've been discussing this for a while - right now the position is to not do any cross-method analysis (magic). The two main reasons are:
* Complexity in the linker - cross-method analysis basically leads to a rather complex multi-pass analysis - so more complex code and very likely quite a bit slower.
* If the auto-propagation leads to a warning - for example let's say we infer this annotation but the callsite violates the annotation - producing a meaningful warning message is very challenging - we would have to somehow trace back the "virtual callstack" of the auto-propagations somehow. The fear is that it would be complex to produce anything remotely usable and that it would end up similar to C++ template errors, which are just a nightmare to read.

I do think that in the future we might want to go in this direction at least a little bit - for example being able to nicely handle async (or any compiler generated state machine) would more or less require this (we've seen this already in some places). We could start with some limited version and so on (for example only do this on private members, but require manual annotation on all public members, ...).

Reply:
Whole program analysis was explicitly scoped out because it has implications on linker throughput.

Plus the potential warning messages that the whole program analysis comes up with would end up being very obscure ("An unknown `System.Type` value was passed through 10 method calls and fields before it reached an API that needs to access a property on it. Please annotate the `System.Type` in that unrelated spot to make the reflection guaranteed work 10 places later.").

These annotations are similar to nullable annotations. The C# compiler could also infer nullable annotations in many places, but doesn't.

Reply:
> Whole program analysis was explicitly scoped out

To be clear, I wasn't actually asking about whole program analysis.  When the linker ran over this assembly as part of the assembly build (not the app build), and when it processed this method, it would add the attribution to this method, without looking at any other methods.

Reply:
Wouldn't it still have to "re-analyze" all callers? (and potentially propagate the annotation another layer above and so on...)

Reply:
> Wouldn't it still have to "re-analyze" all callers? (and potentially propagate the annotation another layer above and so on...)

I don't know how the linker traverses methods today, but the first time it encounters a method, it would depth-first process it in order to update the annotations on the method, prior to analyzing the caller.  There would potentially still need to be some re-analysis in the case of cycles, or it could give up.  The goal wouldn't be to be perfect nor to eliminate the need in cases for the developer to still manually annotate, but it could eliminate some warnings and the need for some manual annotation, just as is the case with the linker's heuristics today.

Reply:
If we do that, for public API it would also need to update the ref assembly so that customers benefit from those attributes, right?